packages/contracts/src/utils/LRUList.sol

```diff
@@ -0,0 +1,132 @@
+            next[addr] = next[head];
+            next[head] = addr;
+        } else {
+            // before: head ->         first  ..  prev -> addr -> another
+            // after:  head -> addr -> first  ..  prev ->         another
+
+            require(next[prevItem] == addr, "given wrong prevItem");
+            address nextHead = next[head];
+            if ( nextHead != addr) {
+                address addrNext = next[addr];
+                next[addr] = nextHead;
+                next[head] = addr;
+                next[prevItem] = addrNext;
+            }
+        }
+    }
+
+    /**
+     * count all items in the list.
+     */
+    function countItems() public view returns (uint ret) {
+        (ret,) = countFrom(address(this), MAX);
+    }
+
+    /**
+     * partial list count (in case the list is insanely large)
+     * if max is reached, then "nextFrom" is set to the value to give "from" on the next call
+     * @param from - start after this item (use "this" to count all)
+     * @param max - maximum count to return
+     */
+    function countFrom(address from, uint max) public view returns (uint ret, address nextFrom){
+        ret = 0;
+        address head = address(this);
+        address p = next[from];
+        require(p != address(0), "from not in list");
+        for (;;) {
+            if (p == head) {
+                return (ret, address(0));
+            }
+            ret++;
+            if (ret >= max) {
+                return (ret, p);
+            }
+            p = next[p];
+        }
+    }
+
+    /**
+     * read all items.
+     */
+    function readAllItems() public view returns (address[] memory ret) {
+        (ret,) = readItemsFrom(address(this), MAX);
+    }
+
+    function readItemsFrom(address from, uint max) public view returns (address[] memory ret, address nextFrom) {
+        uint count;
+        (count, nextFrom) = countFrom(from, max);
+        ret = new address[](count);
+        address pos = from;
+        for (uint i = 0; i < count; i++) {
+            pos = next[pos];
+            ret[i] = pos;
+        }
+    }
+
+    //find item previous to given item.
+    // (works well for normal count. use prevFrom for insanely huge lists)
+    function getPrev(address item) public view returns (address ret) {
+        //check if not in list (instead of revert)
+        if ( next[item]==address(0))
+            return address (0);
+        (ret,) = prevFrom(item, address(this), MAX);
+    }
+
+
+    //scan list in chunks
+    // item - the item to find the predecessor of.
+    // from - the first address to lookup from. initialize to "this" on first call.
+    // scanCount - max calls to attempt
+    // returns: (ret,null) - "ret" is the predecessor of "item"
+    //      (null,nextFrom) - didn't find result. repeat calling this method with (item,nextFrom,count)
+    function prevFrom(address item, address from, uint scanCount) public view returns (address ret, address nextFrom) {
```

Comment:
What is the use case for these two methods (`getPrev` and `prevFrom`)?

Reply:
getPrev(x) is "high level" function, and as long as the list is not insanely large, is good enough.
if someone bombard the list with items so that getPrev fails to read them, then one have to scan in chunks using prevFrom. 
I validated the code to run (in test), but my relayer and client implementations don't use this "chunked" mode.

Reply:
If I understand it correctly, its job is to go backwards one step in a  unidirectional singly linked list, is that so?
This seems to me like an abuse of this data structure. Why is it needed?